src/nbl/asset/utils/CHLSLCompiler.cpp

```diff
@@ -61,29 +92,123 @@ CHLSLCompiler::~CHLSLCompiler()
 }
 
 
-struct DxcCompilationResult
+static void try_upgrade_hlsl_version(std::vector<std::wstring>& arguments, system::logger_opt_ptr& logger)
 {
-    ComPtr<IDxcBlobEncoding> errorMessages;
-    ComPtr<IDxcBlob> objectBlob;
-    ComPtr<IDxcResult> compileResult;
+    auto stageArgumentPos = std::find(arguments.begin(), arguments.end(), L"-HV");
+    if (stageArgumentPos != arguments.end()) {
+        auto index = stageArgumentPos - arguments.begin() + 1; // -HV XXXXX, get index of second
+        std::wstring version = arguments[index];
+        if (!isalpha(version.back()) && version.length() >= 4 && std::stoi(version) < 2021)
+            arguments[index] = L"2021";
+    }
+    else {
+        logger.log("Compile flag error: Required compile flag not found -HV. Force enabling -HV 202x, as it is required by Nabla.", system::ILogger::ELL_WARNING);
+        arguments.push_back(L"-HV");
+        arguments.push_back(L"202x");
+    }
+}
 
-    std::string GetErrorMessagesString()
+
+static void try_upgrade_shader_stage(std::vector<std::wstring>& arguments, asset::IShader::E_SHADER_STAGE shaderStageOverrideFromPragma, system::logger_opt_ptr& logger) {
+	
+	constexpr int MajorReqVersion = 6,
+		MinorReqVersion = 7;
+
+	auto foundShaderStageArgument = std::find(arguments.begin(), arguments.end(), L"-T");
+	if (foundShaderStageArgument != arguments.end() && foundShaderStageArgument + 1 != arguments.end()) {
+		auto foundShaderStageArgumentValueIdx = foundShaderStageArgument - arguments.begin() + 1;
+		std::wstring s = arguments[foundShaderStageArgumentValueIdx];
+        if (s.length() >= 6) {
+            std::vector<std::wstring::iterator> underscorePositions = {};
+            auto it = std::find(s.begin(), s.end(), '_');
+            while (it != s.end()) {
+                underscorePositions.push_back(it);
+                it = std::find(it + 1, s.end(), '_');
+            }
+            if (shaderStageOverrideFromPragma != IShader::ESS_UNKNOWN)
+            {
+                // replace first 2 characters if shaderStageOverrideFromPragma != Unknown
+                auto stageStr = ShaderStageToString(shaderStageOverrideFromPragma);
+                if (!stageStr) {
+                    logger.log("invalid shader stage %i", system::ILogger::ELL_ERROR, shaderStageOverrideFromPragma);
+                    return;
+                }
+                if (underscorePositions.size() == 0)
+                {
+                    logger.log("incorrect format for shader stage parameter, expecting shader stage, underscore, then version", system::ILogger::ELL_ERROR, shaderStageOverrideFromPragma);
+                    return;
+                }
+                int len = underscorePositions[0] - s.begin();
+                s.replace(0, len, stageStr);
```

Comment:
do you even understand what you're doing here?

you could have just deduced the SM version first! then fixed up the stage, you wrote complete sphagetti code